{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "94ea34e1",
   "metadata": {},
   "source": [
    "解析gff3文件，并导出所有gene以及上下游，蛋白，CDS序列，其中基因序列中需要标注出内含子的位置。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8c932d33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current handle is  [<function to_lower at 0x0619C898>, <function func at 0x0619CD18>]\n",
      "now running parsing\n",
      "now creating index for seqlite\n",
      "now read id files\n",
      "now ids is ['FUN_000002-T1\\n', 'FUN_000002-T2\\n', 'FUN_000002-T3']\n",
      "get ids []\n",
      "[]\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import sqlite3\n",
    "import uuid\n",
    " \n",
    " \n",
    "class Type:\n",
    "    gene = 'gene'\n",
    "    mrna = 'mrna'\n",
    "    cds = 'cds'\n",
    "    lnc_rna = 'lnc_rna'\n",
    "    exon = 'exon'\n",
    "    region = 'region'\n",
    "    pseudogene = 'pseudogene'\n",
    "    transcript = 'transcript'\n",
    " \n",
    " \n",
    "def to_lower(x):\n",
    "    return str(x).lower()\n",
    " \n",
    "class GffNode:\n",
    " \n",
    "    id_patt = '\\tID=([\\w.-]+);?'\n",
    "    key_handles = [to_lower,]  # 默认添加一个小写的处理器\n",
    " \n",
    "    def set_node(self, node: str):\n",
    "        self.node = node\n",
    "        self.keys = []\n",
    "        self._nodes = []\n",
    "        return self\n",
    " \n",
    "    def set_key_handle(cls, handle): \n",
    "        cls.key_handles.append(handle)  # 添加处理器,不同的函数\n",
    "        print(\"current handle is \", cls.key_handles)\n",
    " \n",
    "    def do_key_handle(cls, keys: str):\n",
    "        for handle in cls.key_handles: #依次处理\n",
    "            keys = list(map(handle, keys))  # 对提取的id逐个应用处理器\n",
    "        return keys\n",
    " \n",
    "    def _parser_key(self, item: str):\n",
    "        r = re.findall(self.id_patt, item, re.I | re.S)  # 提取id\n",
    "        return r or None \n",
    " \n",
    "    def _set_keys(self, keys: str):\n",
    "        if not keys:\n",
    "            return\n",
    "        keys = self.do_key_handle(keys)\n",
    "        self.keys = set(keys)  # 获取每一个单元中的所有id并去重\n",
    " \n",
    "    def _get_key(self, item: str):\n",
    "        return self.do_key_handle(self._parser_key(item))\n",
    " \n",
    "    def _call_(self):\n",
    "        keys = self._parser_key(self.node)\n",
    "        self._set_keys(keys)\n",
    " \n",
    "    def eq(self, a, b):\n",
    "        if isinstance(a, str) and isinstance(b, str):\n",
    "            return a in b or b in a or a == b\n",
    "        return False\n",
    " \n",
    "    def __contains__(self, key: str):\n",
    "        key = to_lower(key)\n",
    "        return key in self.keys or any([\n",
    "            self.eq(i, key) for i in self.keys\n",
    "        ])\n",
    " \n",
    "    def __len__(self):\n",
    "        return len(self.keys)\n",
    " \n",
    "    def graph(self):\n",
    "        ...\n",
    " \n",
    "    def __str__(self):\n",
    "        return f\"keys - {len(self)} - {self.keys[0]}\"\n",
    " \n",
    "    def parser_node(self):\n",
    "        node = self.node.split('\\n')\n",
    "        uid = str(uuid.uuid4())  # 添加uuid，避免出现同一个gene中出现多条mrna难以区分的情况\n",
    "        for o in node:\n",
    "            o = o.strip()\n",
    "            if not o:\n",
    "                continue\n",
    "            _o = o.split('\\t')\n",
    "            chor = self.safe_lower(_o[0])  # 染色信息\n",
    "            type = self.safe_lower(_o[2])  # 类型信息\n",
    "            start = self.safe_int(_o[3])  # 开始位置\n",
    "            end = self.safe_int(_o[4])  # 结束位置\n",
    "            length = end - start + 1\n",
    "            strand = _o[6]  # 正负链信息\n",
    "            _key = self._get_key(o)  # id\n",
    "            if _key:\n",
    "                key = self.safe_lower(_key[0])\n",
    "            else:\n",
    "                key = 'unknown'\n",
    "            self._nodes.append(dict(zip(\n",
    "                ['chro', 'type', 'start', 'end', 'strand', 'key', 'length','uid','o'],\n",
    "                [chor, type, start, end, strand, key, length,uid,o]  # 保留原始信息o，可以自定义处理\n",
    "            )))\n",
    " \n",
    "    @staticmethod  #静态方法，没有self参数\n",
    "    def safe_lower(o):\n",
    "        if isinstance(o, str):\n",
    "            return o.lower()\n",
    "        return o\n",
    " \n",
    "    @staticmethod\n",
    "    def safe_int(o):\n",
    "        return int(o)   # 其实并不安全。。。\n",
    " \n",
    "    def get(self, *, key: str = None, type=None, all=False):\n",
    "        def func(x): return True\n",
    "        if not self._nodes:\n",
    "            self.parser_node()\n",
    "        if all or (key is None and type is None):\n",
    "            func = lambda x:True\n",
    "        elif key and type:\n",
    "            def func(x):\n",
    "                return self.eq(x['key'].lower(), self.safe_lower(key)) and self.eq(x['type'].lower(), type)\n",
    "        elif key or type:\n",
    "            def func(x):\n",
    "                return self.eq(x['key'].lower(), self.safe_lower(key)) or self.eq(x['type'].lower(), type)\n",
    "        r = list(filter(\n",
    "            func,\n",
    "            self._nodes\n",
    "        ))\n",
    "        return r\n",
    " \n",
    " \n",
    "class Gff:\n",
    "    def __init__(self, gff_file, id_file):\n",
    "        self.gff = open(gff_file, 'r', encoding='utf-8')  #读取gff3文件\n",
    "        self.id_file = open(id_file, 'r', encoding='utf-8') #读取需要提取信息的ID号码\n",
    "        self.index = gff_file + '.index'\n",
    "        self.node = GffNode() #创建GFFnode对象\n",
    " \n",
    "    def run(self):\n",
    "        print(\"now creating index for seqlite\")\n",
    "        self.create_index()  # 建序\n",
    "        print(\"now read id files\")\n",
    "        self.ids = self.read_ids()  # 读取ID文件\n",
    "        print(self.ids)\n",
    " \n",
    "    def set_key_handle(self, handle):\n",
    "        self.node.set_key_handle(handle)\n",
    " \n",
    "    def create_index(self):\n",
    "        import os\n",
    "        if not os.path.exists(self.index):\n",
    "            conn = sqlite3.connect(self.index)\n",
    "            conn.execute('''CREATE TABLE id_index \n",
    "            (id INT  NOT NULL,\n",
    "            id_name TEXT NOT NULL);\n",
    "            ''')\n",
    "            c = conn.cursor()\n",
    "            count = 0\n",
    "            insert = 'INSERT INTO id_index(id,id_name) VALUES '\n",
    "            for node in self.parse(): #创建之前先调用parse获取node信息\n",
    "                node = self.node_call(node)  #为GFFnode实例\n",
    "                sql = ''\n",
    "                for key in node.keys:\n",
    "                    sql += '('\n",
    "                    sql += f'{count},\"{key}\"'\n",
    "                    sql += '),'\n",
    "                count += 1\n",
    "                sql = insert + sql.strip(',')\n",
    "                c.execute(sql)\n",
    "            conn.commit()\n",
    "            self.gff.seek(0)\n",
    "            self.conn = conn\n",
    "        else:\n",
    "            self.conn = sqlite3.connect(self.index)\n",
    " \n",
    "    @staticmethod\n",
    "    def node_call(node):\n",
    "        node._call_()\n",
    "        return node\n",
    " \n",
    "    def parse(self): #\n",
    "        node = ''\n",
    "        for line in self.gff: #读取gff文件每一行\n",
    "            #print(\"current node is \", node, \"end node\")\n",
    "            if not line or '#' in line:\n",
    "                continue\n",
    "            if f'\\t{Type.gene}\\t' in line and node:\n",
    "                #print(\"line is \", line,\"##start noede \", node, '###end node\\n')\n",
    "                yield self.node.set_node(node)\n",
    "                node = ''\n",
    "            node = node + line\n",
    "        yield self.node.set_node(node)\n",
    " \n",
    "    def read_ids(self):\n",
    "        ids = self.id_file.readlines()  #self.id_file是文件句柄\n",
    "        print(\"now ids is\", ids)\n",
    "        ids = [i.replace('\\n', '').lower() for i in ids]\n",
    "        conn = self.conn\n",
    "        _ids = []\n",
    "        for id in ids:\n",
    "            _search_sql = 'select id,id_name from id_index where id_name like \"%s\"' % id\n",
    "            rows = conn.execute(_search_sql)\n",
    "            rows = list(rows)\n",
    "            if rows:\n",
    "                _ids.append(rows[0])\n",
    "        _ids.sort(key=lambda x: x[0], reverse=True)\n",
    "        print('get ids', _ids)\n",
    "        return _ids\n",
    " \n",
    "    def search(self, *, key=None, type=None, all=False, use_strict=True): \n",
    "        \"\"\"\n",
    "            key 即ID\n",
    "            type  是gene、exon、cds等\n",
    "            all  返回整个单元的所有结构信息，优先级最高（此时key或者type会失效）\n",
    "            use_strict  返回结果中是否要通过key在做一次筛选，一般用不到。在多顺反子中可能会用的到。\n",
    "        \n",
    "        \"\"\"\n",
    "        count = 0\n",
    "        ids = []\n",
    "        keys = []\n",
    "        if self.ids:\n",
    "            for _id in self.ids:\n",
    "                if _id[0] not in ids:\n",
    "                    ids.append(_id[0])\n",
    "                keys.append(_id[1])\n",
    "        else:\n",
    "            return []\n",
    "        first = ids.pop()\n",
    "        data = []\n",
    "        for node in self.parse():\n",
    "            if count == first:\n",
    "                node = self.node_call(node)\n",
    "                data = data + node.get(key=key, type=type, all=all)\n",
    "                if not ids:\n",
    "                    break\n",
    "                first = ids.pop()\n",
    "            count += 1\n",
    "        return list(filter(\n",
    "            lambda x: x['key'] in keys if use_strict else True,\n",
    "            data\n",
    "        ))\n",
    " \n",
    " \n",
    "if __name__ == \"__main__\":\n",
    "    p = 'genomic.gff3'\n",
    "    id_file = 'ids.txt'  # 每一行放置一个id\n",
    "    p = Gff(p, id_file) \n",
    "    def func(x):\n",
    "        if 'gene' in x:\n",
    "            return x.split('-')[1]\n",
    "        if 'cds' in x:\n",
    "            return x.split('-', 1)[1]\n",
    "        if 'exon' in x:\n",
    "            return x.split('-')[1]\n",
    "        if 'rna' in x:\n",
    "            return x.split('-', 1)[1]\n",
    "        return x\n",
    "    p.set_key_handle(func)  # 添加自定义的ID处理器  \n",
    "    # 添加处理器的目的是为了将gff3中的id转换成你的id_file中的id\n",
    "    # 所以可以添加多个处理器\n",
    "    # 一步步的进行处理，直到id和id_file中的相同\n",
    "    print(\"now running parsing\")\n",
    "    p.run()\n",
    "    d = p.search(type=Type.gene, use_strict=False)\n",
    "    import json\n",
    "    print(json.dump(d, open('gene.json', 'w')))  # 写入json文件"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
